[AutoFill Debugging] Make several minor adjustments to the text extraction output format

whsieh opened the pull request:
#### df7894a3240074f70d8e3b1ce1b4ea4886c72052
<pre>
[AutoFill Debugging] Make several minor adjustments to the text extraction output format
<a href="https://bugs.webkit.org/show_bug.cgi?id=302352">https://bugs.webkit.org/show_bug.cgi?id=302352</a>
<a href="https://rdar.apple.com/164509017">rdar://164509017</a>

Reviewed by Abrar Rahman Protyasha.

Make several adjustments to text extraction output formatting, and rebaseline / augment tests as
needed. See below for more details.

* LayoutTests/fast/text-extraction/debug-text-extraction-basic-expected.txt:
* LayoutTests/fast/text-extraction/debug-text-extraction-basic.html:
* LayoutTests/fast/text-extraction/debug-text-extraction-ignore-autofilled-fields-expected.txt:
* LayoutTests/fast/text-extraction/debug-text-extraction-lightweight-expected.txt:
* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::extractRecursive):
* Source/WebCore/page/text-extraction/TextExtractionTypes.h:
(WebCore::TextExtraction::Item::dataAs const):

Add a helper method to grab the item&apos;s data variant as the given template type (or `nullopt` if the
type does not match).

* Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
(WebKit::addPartsForItem):
(WebKit::childTextNodeIsRedundant):

Avoid exposing a child text node in the extraction output, in the case where the content of the
child text node is already reflected in one of the attributes on the element (e.g. in the case where
you have `&lt;input aria-label=&quot;foo&quot;&gt;foo&lt;/input&gt;` or `&lt;a href=&quot;<a href="http://example.com&quot">http://example.com&quot</a>;&gt;example.com&lt;/a&gt;`, we
condense that into `input,aria-label=&apos;foo&apos;` and `link,url=&apos;<a href="http://example.com&apos">http://example.com&apos</a>;`, respectively).

(WebKit::addTextRepresentationRecursive):

Instead of exposing a top-level `textFormControl` item type, just use the tag name (`input` or
`textarea`) which more clearly represents the type of the element.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/302898@main">https://commits.webkit.org/302898@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f71daf490c000a4a9767846da6849bbea0865755

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130593 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2864 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41548 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138011 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/82210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132464 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2876 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2756 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/99493 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/82210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133540 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116938 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80200 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/2009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35071 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81270 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/110580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35575 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140491 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2654 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/2410 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/107997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2698 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113283 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107929 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27462 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2049 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/31713 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/55632 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2724 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66113 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/2543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/2745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2650 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
```diff
@@ -469,6 +467,27 @@ static void addPartsForItem(const TextExtraction::Item& item, std::optional<Node
+        if (editable.label.containsIgnoringASCIICase(childText))
+            return true;
+
+        return std::any_of(parent.ariaAttributes.begin(), parent.ariaAttributes.end(), [&](auto& entry) {
```
nit: I am unsure if `HashMap` works with `std::ranges`, but:
```suggestion
        return std::ranges::any_of(parent.ariaAttributes, [&](auto& entry) {
```

whsieh replied:
Yep, that works and is much cleaner — thanks for the review!

whsieh committed:
[AutoFill Debugging] Make several minor adjustments to the text extraction output format
https://bugs.webkit.org/show_bug.cgi?id=302352
rdar://164509017

Reviewed by Abrar Rahman Protyasha.

Make several adjustments to text extraction output formatting, and rebaseline / augment tests as
needed. See below for more details.

* LayoutTests/fast/text-extraction/debug-text-extraction-basic-expected.txt:
* LayoutTests/fast/text-extraction/debug-text-extraction-basic.html:
* LayoutTests/fast/text-extraction/debug-text-extraction-ignore-autofilled-fields-expected.txt:
* LayoutTests/fast/text-extraction/debug-text-extraction-lightweight-expected.txt:
* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::extractRecursive):
* Source/WebCore/page/text-extraction/TextExtractionTypes.h:
(WebCore::TextExtraction::Item::dataAs const):

Add a helper method to grab the item's data variant as the given template type (or `nullopt` if the
type does not match).

* Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
(WebKit::addPartsForItem):
(WebKit::childTextNodeIsRedundant):

Avoid exposing a child text node in the extraction output, in the case where the content of the
child text node is already reflected in one of the attributes on the element (e.g. in the case where
you have `<input aria-label="foo">foo</input>` or `<a href="http://example.com">example.com</a>`, we
condense that into `input,aria-label='foo'` and `link,url='http://example.com'`, respectively).

(WebKit::addTextRepresentationRecursive):

Instead of exposing a top-level `textFormControl` item type, just use the tag name (`input` or
`textarea`) which more clearly represents the type of the element.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/302898@main
